### PR TITLE
add functions relavent of QAOA

### PR DESCRIPTION
**Pull request by FelixXu35**
Hi,

Thanks for reviewing this pull request.
Those are the functions that I (potentially the users) need in QAOA.
I put them into different py files based on my understanding.
Please feel free to let me know if there are things need to add, modify or move.
I will PR the tutorials after those functions go online. Since I need to import the functions in the tutorials.

Thanks.
Felix

### CODE REVIEW

File: tensorcircuit/applications/__init__.py [hunk 1]
```diff
@@ -3,3 +3,5 @@
 the code inside is subject to change, be caution to use.
 Most of the useful code is and will be refactored and copied to other parts of tensorcircuit.
 """
+
+from . import optimization
```
Comment (refraction-ray) [inline tensorcircuit/applications/__init__.py:7]: I dont suggest to explicitly import any submodule in applications folder, but you can still use them later anyway

File: tensorcircuit/applications/optimization.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
+"""
+modules for QUBO problems in QAOA
+"""
+
+from typing import List, Callable, Any
+
+import tensorcircuit as tc
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:7]: please not import tc within tc

File: tensorcircuit/applications/optimization.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+import numpy as np
+from functools import partial
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:9]: this import is a bultin import, should be in the first group

File: tensorcircuit/applications/optimization.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+import tensorflow as tf
+import scipy.optimize as optimize
+
+from ..cons import backend
+from ..templates.blocks import QAOA_ansatz_for_Ising
+from ..templates.conversions import QUBO_to_Ising
+from tensorflow.python.ops.numpy_ops import np_config
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:16]: this should be in the second import group, no comment on group order for later files

File: tensorcircuit/applications/optimization.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+np_config.enable_numpy_behavior()
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:18]: what is this for

File: tensorcircuit/templates/conversions.py [hunk 6]
```diff
@@ -36,3 +38,201 @@ def get_ps(qo: Any, n: int) -> Tuple[Tensor, Tensor]:
         res.append(res_t)
         wts.append(w)
     return np.array(res), np.array(wts)
+
+
+def QUBO_to_Ising(Q: List[list]) -> Tuple[List[list], list, float]:
```
Comment (refraction-ray) [inline tensorcircuit/templates/conversions.py:43]: in the type annotation, explicit label a list of what

**Comment by FelixXu35** [comment]
will do

File: tensorcircuit/templates/conversions.py [hunk 7]
```diff
@@ -36,3 +38,201 @@ def get_ps(qo: Any, n: int) -> Tuple[Tensor, Tensor]:
...
+            )  # Calculate the weight for the two-qubit interaction term
+            weights = np.concatenate(
+                (weights, weight), axis=None
+            )  # Add the weight to the weights list
+
+    return pauli_terms, weights, offset
+
+
+def QUBO_from_portfolio(cov: Array, mean: Array, q: float, B: int, t: float) -> Tensor:
+    """
```
Comment (refraction-ray) [inline tensorcircuit/templates/conversions.py:99]: maybe also move this to applications/finance?

File: tensorcircuit/templates/blocks.py [hunk 8]
```diff
@@ -200,3 +200,97 @@ def qft(
...
+    :param mixer: mixer type. The options are "X", "XY_ring", "XY_par_ring", "XY_full", and "QAMPA".
+    """
+    nqubits = len(pauli_terms[0])
+    c = Circ(nqubits)
+    for i in range(nqubits):
+        c.h(i)  # Apply Hadamard gate to each qubit
+
+    for j in range(nlayers):
+        # cost terms
+        for k in range(len(pauli_terms)):
```
Comment (refraction-ray) [inline tensorcircuit/templates/blocks.py:230]: for k, term in enumerate(pauli_terms):

File: tensorcircuit/templates/blocks.py [hunk 9]
```diff
@@ -200,3 +200,97 @@ def qft(
...
+            term = pauli_terms[k]
+            index_of_ones = []
+            for l in range(len(term)):
+                if term[l] == 1:
+                    index_of_ones.append(l)
+            if len(index_of_ones) == 1:
+                c.rz(index_of_ones[0], theta=2 * weights[k] * params[2 * j])
+                # Apply Rz gate with angle determined by weight and current parameter value
+            elif len(index_of_ones) == 2:
+                c.exp1(
```
Comment (refraction-ray) [inline tensorcircuit/templates/blocks.py:240]: use rzz gate directly instead of exp1 gate

File: tensorcircuit/templates/blocks.py [hunk 10]
```diff
@@ -200,3 +200,97 @@ def qft(
...
+                c.rx(
+                    i, theta=params[2 * j + 1]
+                )  # Apply Rx gate with angle determined by current parameter value
+        # Parity ring XY mixer
+        elif mixer == "XY":
+            pairs = []
+            for index in [0, 1]:
+                while index + 2 <= nqubits:
+                    pairs.append([index, index + 1])
+                    index += 2
```
Comment (refraction-ray) [inline tensorcircuit/templates/blocks.py:262]: write the index using `range` in elegant fashion

File: tensorcircuit/templates/blocks.py [hunk 11]
```diff
@@ -200,3 +200,97 @@ def qft(
...
+            for pair in pairs:
+                c.exp1(pair[0], pair[1], unitary=G._xx_matrix, theta=params[2 * j + 1])
```
Comment (refraction-ray) [inline tensorcircuit/templates/blocks.py:264]: directly use rxx and ryy gate

File: tensorcircuit/templates/blocks.py [hunk 12]
```diff
@@ -200,3 +200,97 @@ def qft(
...
+            for index in [0, 1]:
+                while index + 2 <= nqubits:
+                    pairs.append([index, index + 1])
+                    index += 2
+            for pair in pairs:
+                c.exp1(pair[0], pair[1], unitary=G._zz_matrix, theta=params[2 * j + 1])
+            c.exp1(nqubits - 1, 0, unitary=G._zz_matrix, theta=params[2 * j + 1])
+
+        # ZZ mixer with full couplings
+        elif mixer == "ZZ_full":
```
Comment (refraction-ray) [inline tensorcircuit/templates/blocks.py:288]: better decouple the logic of `XX`, `XY` and `full`, `ring`, the latter is used to generate pairs while the former is used to determine the gate type, in this way the code is DRY (dont repeat yourself)

File: tensorcircuit/applications/optimization.py [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+            params = backend.implicit_randn(shape=[ncircuits, 2 * nlayers], stddev=0.1)
+            # If init_params is not provided, initialize the parameters randomly.
+            # If vvag flag is set to True, use vectorized variational adjoint gradient (vvag) with multiple circuits.
+    else:
+        params = init_params
+        # If init_params is provided, use the provided parameters.
+    # Initialize the parameters for the ansatz circuit.
+
+    if jit == True:
+        loss_val_grad = backend.jit(loss_val_grad, static_argnums=(1, 2))
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:132]: is these any need to switch off jit?

**Comment by refraction-ray** [comment]
please fix mypy issue first as indicated by CI

**Comment by FelixXu35** [comment]
finally passed the CL test. Gonna add tests for new functions

**Comment by FelixXu35** [comment]
✌️Finally passed four checks.
My tests are a little bit tedious. Will try to improve my skills in the future.

File: tensorcircuit/applications/finance/portfolio.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        n_days = [len(i) for i in data]
+        if max(n_days) != (sum(n_days) / len(n_days)):
+            raise Exception("Timespan of stocks should be the same")
+        self.n_days = len(data[1])
+
+        # Calculate the daily percentage price change
+        self.daily_change = []
+        for i in range(self.n_stocks):
+            each_stock = []
+            for j in range(self.n_days - 1):
```
Comment (refraction-ray) [inline tensorcircuit/applications/finance/portfolio.py:66]: np.diff

File: tensorcircuit/applications/finance/portfolio.py [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+        :return: The annualized covariance matrix rounded to the specified number of decimals.
+        """
+        mean = np.mean(self.daily_change, axis=1)
+        relative_change = [
+            [j - mean[i] for j in self.daily_change[i]] for i in range(6)
+        ]
+        cov = 252 / self.n_days * np.dot(relative_change, np.transpose(relative_change))
+        return cov.round(decimals)
+
+    def get_penalty(
```
Comment (refraction-ray) [inline tensorcircuit/applications/finance/portfolio.py:95]: leaving @JAllcock for reviewing this method, not very sure what this method is for

File: tensorcircuit/applications/optimization.py [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+    :param init_params (optional): The initial parameters for the ansatz circuit.
+        Default is None, which initializes the parameters randomly.
+    :paran mixer (optional): The mixer operator to use. Default is "X". The other options are "XY" and "ZZ".
+    :param learning_rate (optional): The learning rate for the Adam optimizer. Default is 1e-2.
+    :param callback (optional): A callback function that is executed during each iteration. Default is None.
+    :param full_coupling (optional): A flag indicating whether to use all-to-all coupling in mixers. Default is False.
+    :return params: The optimized parameters for the ansatz circuit.
+    """
+    if backend != get_backend("tensorflow"):
+        raise ValueError("`QUBO_QAOA` is designed for tensorflow backend.")
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:117]: why jax backend doesn't work? though it is ok that code in applications is backend dependent

File: tensorcircuit/applications/optimization.py [hunk 17]
```diff
@@ -0,0 +1,348 @@
...
+    params: List[float],
+) -> float:
+    """
+    Calculate the CVaR loss for a given QUBO problem using the QAOA ansatz.
+
+    :param nlayers: The number of layers (depth) in the QAOA ansatz.
+    :param Q: The Q-matrix representing the Quadratic Unconstrained Binary Optimization (QUBO) problem.
+    :param nsamples: The number of samples to take for measurements in the CVaR calculation.
+    :param alpha: The cut-off percentage for CVaR.
+    :param fake: A flag indicating the type of CVaR ansatz (circuit-based or expectation-based).
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:279]: fake is not a good name for this option, try measurement_based: bool?

File: tensorcircuit/applications/optimization.py [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+
+def cvar_from_expectation(circuit: Circuit, Q: Tensor, alpha: float) -> float:
+    """
+    Calculate the Conditional Value at Risk (CVaR) from the expectation values of a quantum circuit.
+
+    :param circuit: The quantum circuit.
+    :param Q: The Q-matrix representing the Quadratic Unconstrained Binary Optimization (QUBO) problem.
+    :param alpha: The cut-off percentage for CVaR.
+    :return: The calculated CVaR value.
+    """
```
Comment (refraction-ray) [inline tensorcircuit/applications/optimization.py:242]: if we make this function differentiable (no numpy), the cvar qaoa can be gradient descent, too. Though, in this pr, it is ok to keep thing as it is, maybe you can add some TODO comment

File: tensorcircuit/templates/ansatz.py [hunk 19]
```diff
@@ -1,3 +1,92 @@
...
+    """
+    Construct the QAOA ansatz for the Ising Model.
+    The number of qubits is determined by `pauli_terms`.
+
+    :param params: A list of parameter values used in the QAOA ansatz.
+    :param nlayers: The number of layers in the QAOA ansatz.
+    :pauli_terms: A list of Pauli terms, where each term is represented as a list of 0/1 series.
+    :param weights: A list of weights corresponding to each Pauli term.
+    :param full_coupling (optional): A flag indicating whether to use all-to-all coupling in mixers. Default is False.
+    :paran mixer (optional): The mixer operator to use. Default is "X". The other options are "XY" and "ZZ".
```
Comment (refraction-ray) [inline tensorcircuit/templates/ansatz.py:30]: return?

File: tensorcircuit/applications/optimization.py (hunk 16 above)
Comment (FelixXu35) [reply tensorcircuit/applications/optimization.py:117]: Okay i will just remove the backend check

File: tensorcircuit/applications/finance/portfolio.py (hunk 15 above)
Comment (FelixXu35) [reply tensorcircuit/applications/finance/portfolio.py:95]: https://link.springer.com/article/10.1007/s11128-022-03766-5
This paper introduces a method to determine the ideal penalty, and I included this method.
Although this method needs brutal force and is not capable in a more extensive database, it could indicate the range of the penalty.

**Review by JAllcock — APPROVED** [review]

**Comment by refraction-ray** [comment]
LGTM

**Comment by refraction-ray** [comment]
@all-contributors please add @FelixXu35 for code, tests
